<a href="https://colab.research.google.com/github/xcl-esc/Analise_Goodreads/blob/main/C%C3%B3digos/Limpeza_e_entendimento.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
cd /content/drive/MyDrive/projeto_novo/Códigos/

/content/drive/MyDrive/projeto_novo/Códigos


In [3]:
import pandas as pd
import numpy as np
import os
import pyarrow


In [4]:
path = '/content/drive/MyDrive/projeto_novo/Dados/books.csv'
df = pd.read_csv(path, on_bad_lines='skip')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11123 non-null  int64  
 1   title               11123 non-null  object 
 2   authors             11123 non-null  object 
 3   average_rating      11123 non-null  float64
 4   isbn                11123 non-null  object 
 5   isbn13              11123 non-null  int64  
 6   language_code       11123 non-null  object 
 7     num_pages         11123 non-null  int64  
 8   ratings_count       11123 non-null  int64  
 9   text_reviews_count  11123 non-null  int64  
 10  publication_date    11123 non-null  object 
 11  publisher           11123 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


In [5]:
df.head()

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
0,1,Harry Potter and the Half-Blood Prince (Harry ...,J.K. Rowling/Mary GrandPré,4.57,0439785960,9780439785969,eng,652,2095690,27591,9/16/2006,Scholastic Inc.
1,2,Harry Potter and the Order of the Phoenix (Har...,J.K. Rowling/Mary GrandPré,4.49,0439358078,9780439358071,eng,870,2153167,29221,9/1/2004,Scholastic Inc.
2,4,Harry Potter and the Chamber of Secrets (Harry...,J.K. Rowling,4.42,0439554896,9780439554893,eng,352,6333,244,11/1/2003,Scholastic
3,5,Harry Potter and the Prisoner of Azkaban (Harr...,J.K. Rowling/Mary GrandPré,4.56,043965548X,9780439655484,eng,435,2339585,36325,5/1/2004,Scholastic Inc.
4,8,Harry Potter Boxed Set Books 1-5 (Harry Potte...,J.K. Rowling/Mary GrandPré,4.78,0439682584,9780439682589,eng,2690,41428,164,9/13/2004,Scholastic


Problemas encontrados:

- O rótulo 'num_pages' contém um espaço vazio antes do nome
- rotulo isbn formatado como texto (pq?) - O rótulo isbn é uma identificação do livro, por isso está setado como texto(como um id)
- publication_date precisa passar para date_time
- livros com 0 avaliações mas possuem notas (possível erro de extração, ou correlações errôneas com edições diferentes do mesmo livro entre outros problemas). Não há possibilidade de correção por possuírem vários casos diferentes, logo ficarão como estão.

In [6]:
df.columns = df.columns.str.strip()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11123 non-null  int64  
 1   title               11123 non-null  object 
 2   authors             11123 non-null  object 
 3   average_rating      11123 non-null  float64
 4   isbn                11123 non-null  object 
 5   isbn13              11123 non-null  int64  
 6   language_code       11123 non-null  object 
 7   num_pages           11123 non-null  int64  
 8   ratings_count       11123 non-null  int64  
 9   text_reviews_count  11123 non-null  int64  
 10  publication_date    11123 non-null  object 
 11  publisher           11123 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


In [8]:
df.tail()

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
11118,45631,Expelled from Eden: A William T. Vollmann Reader,William T. Vollmann/Larry McCaffery/Michael He...,4.06,1560254416,9781560254416,eng,512,156,20,12/21/2004,Da Capo Press
11119,45633,You Bright and Risen Angels,William T. Vollmann,4.08,0140110879,9780140110876,eng,635,783,56,12/1/1988,Penguin Books
11120,45634,The Ice-Shirt (Seven Dreams #1),William T. Vollmann,3.96,0140131965,9780140131963,eng,415,820,95,8/1/1993,Penguin Books
11121,45639,Poor People,William T. Vollmann,3.72,0060878827,9780060878825,eng,434,769,139,2/27/2007,Ecco
11122,45641,Las aventuras de Tom Sawyer,Mark Twain,3.91,8497646983,9788497646987,spa,272,113,12,5/28/2006,Edimat Libros


In [9]:
#checagem pré-conversão
print(df.loc[[8177, 11094], ['publication_date', 'title']])

      publication_date                                              title
8177        11/31/2000  In Pursuit of the Proper Sinner (Inspector Lyn...
11094        6/31/1982         Montaillou  village occitan de 1294 à 1324


In [10]:
df['publication_date'] = pd.to_datetime(df['publication_date'], errors='coerce')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   bookID              11123 non-null  int64         
 1   title               11123 non-null  object        
 2   authors             11123 non-null  object        
 3   average_rating      11123 non-null  float64       
 4   isbn                11123 non-null  object        
 5   isbn13              11123 non-null  int64         
 6   language_code       11123 non-null  object        
 7   num_pages           11123 non-null  int64         
 8   ratings_count       11123 non-null  int64         
 9   text_reviews_count  11123 non-null  int64         
 10  publication_date    11121 non-null  datetime64[ns]
 11  publisher           11123 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(5), object(5)
memory usage: 1.0+ MB


In [12]:
df['publication_date'].isna().sum()

np.int64(2)

Precisamos ver quem são as linhas com datas nulas pra decidir se devem ser excluídas ou não.

In [13]:
checagem = df[df['publication_date'].isna()]
checagem

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
8177,31373,In Pursuit of the Proper Sinner (Inspector Lyn...,Elizabeth George,4.10,0553575104,9780553575101,eng,718,10608,295,NaT,Bantam Books
11094,45531,Montaillou village occitan de 1294 à 1324,Emmanuel Le Roy Ladurie/Emmanuel Le Roy-Ladurie,3.96,2070323285,9782070323289,fre,640,15,2,NaT,Folio histoire


Essas duas linhas:
- 8177        11/31/2000  In Pursuit of the Proper Sinner (Inspector Lyn...)
- 11094        6/31/1982         Montaillou  village occitan de 1294 à 1324

Possuem erros de preenchimento de datas, no primeiro e no segundo caso (nos meses 11 e 6) o mes possui apenas 30 dias, e em ambos está preenchido com dia ou mes errados. Por esse motivo não há como tratar essas linhas por não haver informações que respaldem qual o erro com precisão e essas linhas serão excluídas do dataset. O impacto é quase nenhum considerando que representam 0,017% do dataset.

In [14]:
df.dropna(subset=['publication_date'], inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11121 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   bookID              11121 non-null  int64         
 1   title               11121 non-null  object        
 2   authors             11121 non-null  object        
 3   average_rating      11121 non-null  float64       
 4   isbn                11121 non-null  object        
 5   isbn13              11121 non-null  int64         
 6   language_code       11121 non-null  object        
 7   num_pages           11121 non-null  int64         
 8   ratings_count       11121 non-null  int64         
 9   text_reviews_count  11121 non-null  int64         
 10  publication_date    11121 non-null  datetime64[ns]
 11  publisher           11121 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(5), object(5)
memory usage: 1.1+ MB


Abaixo é possível ver livros com notas de avalição mesmo não possuindo nenhuma avaliação.

In [15]:
filtro_nota_erro = (df['average_rating'] > 0) & (df['ratings_count'] < 1)
resultado_nota = df[filtro_nota_erro]
resultado_nota.head(20)

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
264,797,Lonely Planet Londres,Lonely Planet/Sarah Johnstone/Tom Masters,4.03,8408064762,9788408064763,spa,480,0,0,2006-05-01,Geoplaneta
525,1658,American Government: Continuity and Change Al...,Karen O'Connor/Larry J. Sabato,2.83,0321317106,9780321317100,eng,664,0,0,2005-03-11,Longman Publishing Group
526,1664,Essentials of American and Texas Government: C...,Karen O'Connor/Larry J. Sabato,3.50,0321365208,9780321365200,eng,854,0,0,2005-07-29,Longman Publishing Group
624,2034,Comoediae 1: Acharenses/Equites/Nubes/Vespae/P...,Aristophanes/F.W. Hall/W.M. Geldart,5.00,0198145047,9780198145042,grc,364,0,0,1922-02-22,Oxford University Press USA
747,2411,Melville and the politics of identity: From *K...,Julian Markels,3.33,0252063023,9780252063022,eng,164,0,0,1993-07-01,University of Illinois Press
935,3084,April May und June,Elizabeth von Arnim,3.88,345833422X,9783458334224,ger,88,0,0,1995-04-01,Insel Frankfurt
1110,3754,Dr No / Moonraker / Thunderball / From Russia ...,Ian Fleming,3.98,0862731585,9780862731588,eng,862,0,0,1984-01-01,Heinemann-Octopus
1681,5863,V.S. Naipaul,Bruce Alvin King,2.00,1403904561,9781403904560,eng,240,0,0,2003-10-10,Palgrave Macmillan
2051,7400,The Baby Emergency (Tennengarrah Clinic #1),Carol Marinelli,3.60,0263181499,9780263181494,eng,285,0,0,2004-06-01,Mills & Boon
2255,8077,Animales No Se Visten Los (Animals Should Def...,Judi Barrett/Ron Barrett,4.11,1595191356,9781595191359,eng,32,0,0,1991-10-30,Live Oak Media


In [16]:
resultado_nota.count()

,0
bookID,55
title,55
authors,55
average_rating,55
isbn,55
isbn13,55
language_code,55
num_pages,55
ratings_count,55
text_reviews_count,55


In [17]:
df.sort_values(by='ratings_count', ascending=False).head(20)

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
10336,41865,Twilight (Twilight #1),Stephenie Meyer,3.59,0316015849,9780316015844,eng,501,4597666,94265,2006-09-06,Little Brown and Company
1697,5907,The Hobbit or There and Back Again,J.R.R. Tolkien,4.27,0618260307,9780618260300,eng,366,2530894,32871,2002-08-15,Houghton Mifflin
1462,5107,The Catcher in the Rye,J.D. Salinger,3.80,0316769177,9780316769174,eng,277,2457092,43499,2001-01-30,Back Bay Books
307,960,Angels & Demons (Robert Langdon #1),Dan Brown,3.89,1416524797,9781416524793,eng,736,2418736,21303,2006-04-01,Pocket Books
3,5,Harry Potter and the Prisoner of Azkaban (Harr...,J.K. Rowling/Mary GrandPré,4.56,043965548X,9780439655484,eng,435,2339585,36325,2004-05-01,Scholastic Inc.
4415,15881,Harry Potter and the Chamber of Secrets (Harry...,J.K. Rowling/Mary GrandPré,4.42,0439064864,9780439064866,eng,341,2293963,34692,1999-06-02,Arthur A. Levine Books / Scholastic Inc.
1,2,Harry Potter and the Order of the Phoenix (Har...,J.K. Rowling/Mary GrandPré,4.49,0439358078,9780439358071,eng,870,2153167,29221,2004-09-01,Scholastic Inc.
23,34,The Fellowship of the Ring (The Lord of the Ri...,J.R.R. Tolkien,4.36,0618346252,9780618346257,eng,398,2128944,13670,2003-09-05,Houghton Mifflin Harcourt
2114,7613,Animal Farm,George Orwell/Boris Grabnar/Peter Škerl,3.93,0452284244,9780452284241,eng,122,2111750,29677,2003-05-06,NAL
0,1,Harry Potter and the Half-Blood Prince (Harry ...,J.K. Rowling/Mary GrandPré,4.57,0439785960,9780439785969,eng,652,2095690,27591,2006-09-16,Scholastic Inc.


In [18]:
df.sort_values(by='publication_date', ascending=False).head(10)

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
9664,38568,A Quick Bite (Argeneau #1),Lynsay Sands,3.91,0060773758,9780060773755,eng,360,35275,1370,2020-03-31,Avon
10335,41864,American Genesis: Captain John Smith and the F...,Alden T. Vaughan,3.43,0673393550,9780673393555,eng,224,23,2,2019-08-17,Pearson
3981,14142,The Art of Loving,Erich Fromm/Peter D. Kramer/Rainer Funk,4.04,0061129739,9780061129735,eng,192,38148,1310,2019-08-06,Harper Perennial Modern Classics
10769,43940,Object Thinking,David West,3.99,0735619654,9780735619654,eng,334,155,21,2019-07-23,Microsoft Press
1070,3638,The Wish Giver: Three Tales of Coven Tree,Bill Brittain/Andrew Glass,3.83,0064401685,9780064401685,eng,192,2201,218,2019-04-02,HarperCollins
10823,44184,Monster,Walter Dean Myers,3.72,0064407314,9780064407311,eng,281,56171,5341,2019-03-05,Amistad
390,1337,Enthusiasm and Divine Madness,Josef Pieper/Richard Winston/Clara Winston,4.50,189031823X,9781890318239,eng,125,26,4,2019-02-11,St. Augustines Press
4082,14551,The Perfume Factory,Alex Austin,4.18,1790877792,9781790877799,eng,227,9,2,2018-12-12,Kindle
11085,45495,El alquimista: una fábula para seguir tus sueños,Paulo Coelho/Juan Godó Costa,3.86,0062511408,9780062511409,eng,192,775,43,2018-10-23,Rayo
8068,31016,El diablo de la botella,Robert Louis Stevenson/Diana Castellanos/Eleon...,3.74,9580408556,9789580408550,spa,77,34,2,2018-07-15,Grupo Editorial Norma S.A.


In [19]:
df.sort_values(by='publication_date', ascending=True).head(20)

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
9371,37134,Consider the Lilies,Iain Crichton Smith/Isobel Murray,3.88,0753812932,9780753812938,eng,144,332,33,1900-01-01,Polygon
6488,24459,On Duties (De Officiis),Marcus Tullius Cicero/Walter Miller,3.96,0674990331,9780674990333,eng,448,111,11,1913-01-01,Harvard University Press
6816,25692,Agricola / Germania / Dialogue on Oratory,Tacitus/Maurice Hutton/Robert Maxwell Ogilvie/...,3.95,0674990390,9780674990395,mul,384,87,7,1914-01-01,Harvard University Press
435,1440,History of the Peloponnesian War: Bk. 1-2,Thucydides/C.F. Smith,4.32,0674991206,9780674991200,mul,496,208,7,1919-01-01,Harvard University Press
7141,27411,The Library 1 Books 1-3.9,Apollodorus/James George Frazer,3.93,0674991354,9780674991354,mul,464,21,2,1921-01-01,Harvard University Press (Cambridge MA)/Wm He...
437,1444,History of the Peloponnesian War: Bk. 5-6,Thucydides/C.F. Smith,4.35,0674991222,9780674991224,mul,400,38,0,1921-01-01,Harvard University Press
624,2034,Comoediae 1: Acharenses/Equites/Nubes/Vespae/P...,Aristophanes/F.W. Hall/W.M. Geldart,5.00,0198145047,9780198145042,grc,364,0,0,1922-02-22,Oxford University Press USA
440,1459,History of the Peloponnesian War Bk. 7-8,Thucydides/C.F. Smith,4.29,0674991877,9780674991873,eng,480,35,0,1923-01-01,Harvard University Press
6970,26384,Scripta Minora: Hiero/Agesilaus/Constitution o...,Xenophon/Edgar C. Marchant/G.W. Bowersock,4.00,0674992024,9780674992023,en-US,576,19,2,1925-01-01,Harvard University Press
4001,14235,The Great Gatsby,F. Scott Fitzgerald,3.91,0891906797,9780891906797,eng,182,149,7,1925-09-01,Amereon Ltd


In [20]:
df['language_code'].unique()

array(['eng', 'en-US', 'fre', 'spa', 'en-GB', 'mul', 'grc', 'enm',
       'en-CA', 'ger', 'jpn', 'ara', 'nl', 'zho', 'lat', 'por', 'srp',
       'ita', 'rus', 'msa', 'glg', 'wel', 'swe', 'nor', 'tur', 'gla',
       'ale'], dtype=object)

In [21]:
caminho_limpos = '/content/drive/MyDrive/projeto_novo/Dados/Dados_limpos/'

if not os.path.exists(caminho_limpos):
    os.makedirs(caminho_limpos)

df.to_csv(caminho_limpos + 'dados_limpos.csv', index=False)


print(f"Sucesso! Arquivos salvos em: {caminho_limpos}")

Sucesso! Arquivos salvos em: /content/drive/MyDrive/projeto_novo/Dados/Dados_limpos/
